onnx/defs/nn/defs.cc

```diff
@@ -2829,4 +2829,461 @@ ONNX_OPERATOR_SET_SCHEMA(
+      where D is the dimension of normalized_shape. For example, if normalized_shape is (3, 5) (a 2-dimensional shape),
+      the rms norm is computed over the last 2 dimensions of the input. The computation required by standardization can be
+      described by the following equations.
+      ```
+      XSquared = Mul(X, X)
+      XSquaredMean = ReduceMean<axes=normalized_axes>(XSquared)
+      RMS = Sqrt(XSquaredMean)
+      RMSEps = Add(RMS, epsilon)
+      SqrtRMS = Sqrt(RMSEps)
+      Normalized = Div(X, SqrtRMS)
+      ```
+      where `normalized_axes` is `[axis, ..., rank of X - 1]`. The variables `RMS` stand for root mean square,
+      Depending on `stash_type` attribute, the actual computation
+      must happen in different floating-point precision.
+      For example, if `stash_type` is 1, this operator casts
+      all input variables to 32-bit float, perform the computation, and
+      finally cast `Normalized` back to the original type of `X`.
+      The second stage then scales the outcome of the first stage using:
+      ```
+      Y= Mul(Normalized, Scale)
+      ```
+      Let `d[i]` indicate the i-th dimension of `X`.
+      If `X`'s shape is `[d[0], ..., d[axis-1], d[axis], ..., d[rank-1]]`,
+      the shape of `RMS` is `[d[0], ..., d[axis-1], 1, ..., 1]`.
+      `Y` and `X` have the same shape. This operator supports unidirectional broadcasting
+      (tensors `Scale` and `B` should be unidirectional broadcastable to tensor `X`);
+      for more details please check [the doc](Broadcasting.md).
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    RMSNormalization,
+    23,
+    OpSchema()
+        .SetDoc(RMSNormalization_ver23_doc)
+        .Attr(
+            "axis",
+            "The first normalization dimension: normalization will be performed along dimensions axis : rank(inputs).",
+            AttributeProto::INT, static_cast<int64_t>(-1))
+        .Attr(
+            "epsilon",
+            "The epsilon value to use to avoid division by zero.",
+            AttributeProto::FLOAT, 1e-5f)
+        .Attr(
+            "stash_type",
+            "The floating-point precision used in stage one of the computation.",
+            AttributeProto::INT,
+            static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT))
+        .AllowUncheckedAttributes()
+        .Input(0,
+               "X",
+               "The output of the layer for which the skip connection is being created. "
+               "In general, the shape is (N, C, D1, D2, ... , Dn) for n-dimensional data, where "
+               "D1 to Dn are the spatial dimension sizes and N is the batch size, C is the number of channels. "
+               "The root mean squared norm is taken over the last D dimensions, D is determined by the axis attribute.",
+               "T")
+        .Input(1,
+               "scale",
+               "Scale tensor. Shape is the normalized shape ([axis, .., Dn]) or a scalar (which will be broadcasted to "
```

Comment:
Is it required to be strictly one of these two, or can it be anything that can be broadcast to the [axis, ..., Dn] shape? I am guessing that for an implementation, supporting anything that can be broadcast to that shape should be not much more complex than supporting a scalar (especially when the backend supports broadcast for other ops). And it may be useful to support this generalization in some situations ... eg., if it is of shape [Dn] for example